src/main/java/eu/dissco/orchestration/backend/service/MachineAnnotationServiceService.java

```diff
@@ -503,22 +502,75 @@ private void rollbackToPreviousVersion(MachineAnnotationService currentMas,
     }
   }
 
-  public void deleteMachineAnnotationService(String id, String userId)
+  public void tombstoneMachineAnnotationService(String id, Agent agent)
       throws NotFoundException, ProcessingFailedException {
     var currentMasOptional = repository.getActiveMachineAnnotationService(id);
     if (currentMasOptional.isPresent()) {
       var mas = currentMasOptional.get();
-      mas.setOdsTombstoneMetadata(buildTombstoneMetadata(userId,
-          "Machine Annotation Service tombstoned by user through the orchestration backend."));
-      mas.setOdsStatus(OdsStatus.ODS_TOMBSTONE);
-      repository.deleteMachineAnnotationService(id,
-          mas.getOdsTombstoneMetadata().getOdsTombstoneDate());
-      deleteDeployment(currentMasOptional.get());
+      deleteDeployment(mas);
+      tombstoneHandle(id);
+      var timestamp = Instant.now();
+      var tombstoneMas = buildTombstoneMachineAnnotationService(mas, agent, timestamp);
+      repository.tombstoneMachineAnnotationService(tombstoneMas, timestamp);
+      try {
+        kafkaPublisherService.publishTombstoneEvent(mapper.valueToTree(tombstoneMas), mapper.valueToTree(mas));
+      } catch (JsonProcessingException e){
+        log.error("Unable to publish tombstone event to provenance service", e);
+        throw new ProcessingFailedException("Unable to publish tombstone event to provenance service", e);
+      }
     } else {
       throw new NotFoundException("Requested machine annotation system: " + id + "does not exist");
     }
   }
 
+  private void tombstoneHandle(String handle) throws ProcessingFailedException {
+    var request = fdoRecordService.buildTombstoneRequest(ObjectType.MAS, handle);
+    try {
+      handleComponent.tombstoneHandle(request, handle);
+    } catch (PidException e){
+      log.error("Unable to tombstone handle {}", handle, e);
+      throw new ProcessingFailedException("Unable to tombstone handle", e);
+    }
+  }
+
+  private static MachineAnnotationService buildTombstoneMachineAnnotationService(
```

Comment:
Couldn't we reuse the existing function?

Reply:
Not sure which function you mean here. If you mean `buildTombstoneMetadata()` in `TombstoneUtils`, that's already in there. 

If you mean use the same function in DataMapping, MAS, SourceSystem, I don't think so. In order to create the update patch for the CUT event, we need to make a deep copy of the original. It's a bit verbose, but we need to copy every term from the MAS/SourceSystem/DataMapping into a new version of the object and add the tombstone changes to it (same as we did with Annotatons). Because we're accessing each field of these objects, you can't really generalize the function.

Reply:
Sorry, I believe we already have a function for the deep copy (for new and updated). Ah but here we need to deepcopy the original. Hmm feels a bit weird as it is only for the patch right? Anyway not a major issue, we don't do that much and the k8s calls are slowish anyway